tests/test_device/test_ipu/test_ipu_utils.py

```diff
@@ -0,0 +1,181 @@
+            cfg, 'BACKBONES')
+
+    # non-registered class
+    with pytest.raises(KeyError):
+        cfg = dict(type='VGG')
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES)
+
+    # default_args must be a dict or None
+    with pytest.raises(TypeError):
+        cfg = dict(type='ResNet', depth=50)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES, default_args=1)
+
+    # cfg['type'] should be a str or class
+    with pytest.raises(TypeError):
+        cfg = dict(type=1000)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES)
+
+    # cfg should contain the key "type"
+    with pytest.raises(KeyError, match='must contain the key "type"'):
+        cfg = dict(depth=50, stages=4)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES)
+
+    # cfg or default_args should contain the key "type"
+    with pytest.raises(KeyError, match='must contain the key "type"'):
+        cfg = dict(depth=50)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES, default_args=dict(stages=4))
+
+    # incorrect registry type
+    with pytest.raises(TypeError):
+        cfg = dict(type='ResNet', depth=50)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, 'BACKBONES')
+
+    # incorrect default_args type
+    with pytest.raises(TypeError):
+        cfg = dict(type='ResNet', depth=50)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES, default_args=0)
+
+    # incorrect arguments
+    with pytest.raises(TypeError):
+        cfg = dict(type='ResNet', non_existing_arg=50)
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES)
+
+    # cfg not dict
+    with pytest.raises(TypeError):
+        cfg = []
+        model = mmcv.device.ipu.utils.build_from_cfg_with_wrapper(
+            cfg, BACKBONES)
+
+
+@skip_no_ipu
+def test_parse_ipu_options():
+    options_cfg = dict(
+        randomSeed=888,
+        enableExecutableCaching='cache_engine',
+        train_cfg=dict(
+            executionStrategy='SameAsIpu',
+            Training=dict(gradientAccumulation=8),
+            availableMemoryProportion=[0.3, 0.3, 0.3, 0.3],
+        ),
+        eval_cfg=dict(deviceIterations=1, ),
+    )
+    cast_to_options(copy.deepcopy(options_cfg))
+
+    with pytest.raises(NotImplementedError, match='cfg type'):
+        _options_cfg = copy.deepcopy(options_cfg)
+        _options_cfg['randomSeed'] = (1, 3)
+        cast_to_options(_options_cfg)
+
+    with pytest.raises(NotImplementedError, match='options_node type'):
+        _options_cfg = copy.deepcopy(options_cfg)
+        _options_cfg['train_cfg']['Precision'] = {'autocast_policy': 123}
+        cast_to_options(_options_cfg)
+
```

Comment:
Is it possible to test `recomputation_checkpoint`?

Reply:
can only check by profile manully

Reply:
ok